toolkit/tools/pkgworker/pkgworker.go

```diff
@@ -177,9 +179,23 @@ func buildSRPMInChroot(chrootDir, rpmDirPath, toolchainDirPath, workerTar, srpmF
 		return
 	}
 
-	err = chroot.Run(func() (err error) {
-		return buildRPMFromSRPMInChroot(srpmFileInChroot, outArch, runCheck, defines, packagesToInstall, useCcache)
-	})
+	// Run the build in a go routine so we can monitor and kill it if it takes too long.
+	results := make(chan error)
+	go func() {
+		buildErr := chroot.Run(func() (err error) {
+			return buildRPMFromSRPMInChroot(srpmFileInChroot, outArch, runCheck, defines, packagesToInstall, useCcache)
+		})
+		results <- buildErr
+	}()
+
+	select {
+	case err = <-results:
+	case <-time.After(timeout):
+		shell.PermanentlyStopAllProcesses(unix.SIGKILL)
+		logger.Log.Errorf("Timeout after %v: killing all processes in chroot...", timeout)
+		err = fmt.Errorf("build timed out after %s", timeout)
```

Comment:
Can we either depend on the top-level caller to print the error or print it only here? It seems like we can end up with a duplicate error message.

Reply:
The top-level error gets delayed a lot by the cleanup flows. (~30 lines too late). Having this hear makes it clear why we suddenly stopped doing normal build stuff and started running `kill` and `unmount` etc. I'd like to leave it.